src/array.jl

```diff
@@ -331,6 +331,12 @@ setindex!{T}(A::CatArray, v::CategoricalValue{T}, i::Int) =
 # Method preserving levels and more efficient than AbstractArray one
 copy(A::CatArray) = deepcopy(A)
 
+function copy!(A::CatArray, B::CatArray)
+    levels!(A, levels(B))
+    copy!(A.refs, B.refs)
```

Comment:
Better call `copy!` first, so that if it fails during bounds checking, the original array isn't corrupt.

Also, you cannot call `levels!(A, ...)` like that: it will fail if dropping used levels. Calls `levels!(A.pool, ...)` instead. Finally, there's going to be a problem if levels are not in the same order as index; to avoid that, the simplest solution is to call `levels!(A.pool, [])` before calling it with the new levels: that way, you're certain levels and index will be the same.